--- a/drivers/net/wireless/xradio/hwio.c
+++ b/drivers/net/wireless/xradio/hwio.c
@@ -75,7 +75,7 @@
 static inline int __xradio_read_reg32(struct xradio_common *hw_priv,
                                        u16 addr, u32 *val)
 {
-	return __xradio_read(hw_priv, addr, val, sizeof(val), 0);
+	return __xradio_read(hw_priv, addr, val, sizeof(*val), 0);
 }
 
 static inline int __xradio_write_reg32(struct xradio_common *hw_priv,
--- a/drivers/net/wireless/xradio/hwio.h
+++ b/drivers/net/wireless/xradio/hwio.h
@@ -179,7 +179,7 @@
 static inline int xradio_reg_read_32(struct xradio_common *hw_priv,
                                       u16 addr, u32 *val)
 {
-	return xradio_reg_read(hw_priv, addr, val, sizeof(val));
+	return xradio_reg_read(hw_priv, addr, val, sizeof(*val));
 }
 
 static inline int xradio_reg_write_32(struct xradio_common *hw_priv,
@@ -205,7 +205,7 @@
 static inline int xradio_apb_read_32(struct xradio_common *hw_priv,
                                       u32 addr, u32 *val)
 {
-	return xradio_apb_read(hw_priv, addr, val, sizeof(val));
+	return xradio_apb_read(hw_priv, addr, val, sizeof(*val));
 }
 
 static inline int xradio_apb_write_32(struct xradio_common *hw_priv,
@@ -217,7 +217,7 @@
 static inline int xradio_ahb_read_32(struct xradio_common *hw_priv,
                                       u32 addr, u32 *val)
 {
-	return xradio_ahb_read(hw_priv, addr, val, sizeof(val));
+	return xradio_ahb_read(hw_priv, addr, val, sizeof(*val));
 }
 
 static inline int xradio_ahb_write_32(struct xradio_common *hw_priv,
