--- a/drivers/clk/sunxi-ng/ccu-sun50i-h616.c
+++ b/drivers/clk/sunxi-ng/ccu-sun50i-h616.c
@@ -1030,7 +1030,7 @@
 	.num = CLK_NUMBER,
 };
 
-static const struct ccu_reset_map sun50i_h616_ccu_resets[] = {
+static struct ccu_reset_map sun50i_h616_ccu_resets[] = {
 	[RST_MBUS]		= { 0x540, BIT(30) },
 
 	[RST_BUS_DE]		= { 0x60c, BIT(16) },
--- a/drivers/clk/sunxi-ng/ccu-sun8i-de2.c
+++ b/drivers/clk/sunxi-ng/ccu-sun8i-de2.c
@@ -176,7 +176,7 @@
 	.num	= CLK_NUMBER_WITHOUT_ROT,
 };
 
-static const struct ccu_reset_map sun8i_a83t_de2_resets[] = {
+static struct ccu_reset_map sun8i_a83t_de2_resets[] = {
 	[RST_MIXER0]	= { 0x08, BIT(0) },
 	/*
 	 * Mixer1 reset line is shared with wb, so only RST_WB is
@@ -186,7 +186,7 @@
 	[RST_ROT]	= { 0x08, BIT(3) },
 };
 
-static const struct ccu_reset_map sun8i_h3_de2_resets[] = {
+static struct ccu_reset_map sun8i_h3_de2_resets[] = {
 	[RST_MIXER0]	= { 0x08, BIT(0) },
 	/*
 	 * Mixer1 reset line is shared with wb, so only RST_WB is
@@ -196,20 +196,20 @@
 	[RST_WB]	= { 0x08, BIT(2) },
 };
 
-static const struct ccu_reset_map sun50i_a64_de2_resets[] = {
+static struct ccu_reset_map sun50i_a64_de2_resets[] = {
 	[RST_MIXER0]	= { 0x08, BIT(0) },
 	[RST_MIXER1]	= { 0x08, BIT(1) },
 	[RST_WB]	= { 0x08, BIT(2) },
 	[RST_ROT]	= { 0x08, BIT(3) },
 };
 
-static const struct ccu_reset_map sun50i_h5_de2_resets[] = {
+static struct ccu_reset_map sun50i_h5_de2_resets[] = {
 	[RST_MIXER0]	= { 0x08, BIT(0) },
 	[RST_MIXER1]	= { 0x08, BIT(1) },
 	[RST_WB]	= { 0x08, BIT(2) },
 };
 
-static const struct ccu_reset_map sun50i_h616_de33_resets[] = {
+static struct ccu_reset_map sun50i_h616_de33_resets[] = {
 	[RST_MIXER0]	= { 0x00, BIT(0) },
 	[RST_MIXER1]	= { 0x00, BIT(1) },
 	[RST_WB]	= { 0x00, BIT(4) },
